presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java

```diff
@@ -0,0 +1,261 @@
+            case Types.VARCHAR:
+            case Types.NVARCHAR:
+            case Types.LONGVARCHAR:
+            case Types.LONGNVARCHAR:
+                if (columnSize == 0 || columnSize > VarcharType.MAX_LENGTH) {
+                    return createUnboundedVarcharType();
+                }
+                return createVarcharType(columnSize);
+            case Types.BINARY:
+            case Types.VARBINARY:
+            case Types.LONGVARBINARY:
+                return VARBINARY;
+            case Types.DATE:
+                return DATE;
+            case Types.TIME:
+                return TIME;
+            case Types.TIMESTAMP:
+                return TIMESTAMP;
+            case Types.ARRAY:
+                PDataType<?> baseType = PDataType.fromTypeId(rawTypeId - PDataType.ARRAY_TYPE_BASE);
+                Type basePrestoType = toPrestoType(baseType.getSqlType(), columnSize, decimalDigits, arraySize, rawTypeId);
+                return new ArrayType(basePrestoType);
+        }
+        return null;
+    }
+
+    public static Object getObjectValue(Type type, Block block, int position)
+            throws SQLException
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        if (BOOLEAN.equals(type)) {
+            return type.getBoolean(block, position);
+        }
+        else if (BIGINT.equals(type)) {
+            return type.getLong(block, position);
+        }
+        else if (INTEGER.equals(type)) {
+            return toIntExact(type.getLong(block, position));
+        }
+        else if (SMALLINT.equals(type)) {
+            return Shorts.checkedCast(type.getLong(block, position));
+        }
+        else if (TINYINT.equals(type)) {
+            return SignedBytes.checkedCast(type.getLong(block, position));
+        }
+        else if (DOUBLE.equals(type)) {
+            return type.getDouble(block, position);
+        }
+        else if (REAL.equals(type)) {
+            return intBitsToFloat(toIntExact(type.getLong(block, position)));
+        }
+        else if (type instanceof DecimalType) {
+            return readBigDecimal((DecimalType) type, block, position);
+        }
+        else if (isVarcharType(type) || isCharType(type)) {
+            return type.getSlice(block, position).toStringUtf8();
+        }
+        else if (VARBINARY.equals(type)) {
+            return type.getSlice(block, position).getBytes();
+        }
+        else if (DATE.equals(type)) {
+            // convert to midnight in default time zone
+            long utcMillis = DAYS.toMillis(type.getLong(block, position));
+            long localMillis = getInstanceUTC().getZone().getMillisKeepLocal(DateTimeZone.getDefault(), utcMillis);
+            return new Date(localMillis);
+        }
+        else if (TIMESTAMP.equals(type)) {
+            long millisUtc = type.getLong(block, position);
+            return new Timestamp(millisUtc);
```

Comment:
this code should check `io.prestosql.spi.connector.ConnectorSession#isLegacyTimestamp`

This code is not correct  regardless of TIMESTAMP semantics
(unless JVM zone is UTC or equal to presto user session zone)